Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp

```diff
@@ -661,6 +661,78 @@ Protocol::ErrorStringOr<void> InspectorDebuggerAgent::removeBreakpoint(const Pro
     return { };
 }
 
+static String getFunctionName(JSC::CodeBlock* codeBlock)
+{
+    if (!codeBlock)
+        return nullString();
+
+    if (codeBlock->codeType() != JSC::CodeType::FunctionCode)
+        return nullString();
+
+    return JSC::jsCast<JSC::FunctionExecutable*>(codeBlock->ownerExecutable())->ecmaName().string();
+}
+
+Protocol::ErrorStringOr<void> InspectorDebuggerAgent::addSymbolicBreakpoint(const String& symbol, std::optional<bool>&& caseSensitive, std::optional<bool>&& isRegex, RefPtr<JSON::Object>&& options)
+{
+    Protocol::ErrorString errorString;
+
+    auto breakpoint = debuggerBreakpointFromPayload(errorString, WTFMove(options));
+    if (!breakpoint)
+        return makeUnexpected(errorString);
+
+    {
+        SymbolicBreakpoint symbolicBreakpoint;
+        symbolicBreakpoint.symbol = symbol;
+        if (caseSensitive)
+            symbolicBreakpoint.caseSensitive = *caseSensitive;
+        if (isRegex)
+            symbolicBreakpoint.isRegex = *isRegex;
+        symbolicBreakpoint.specialBreakpoint = WTFMove(breakpoint);
+
+        if (!m_symbolicBreakpoints.appendIfNotContains(WTFMove(symbolicBreakpoint)))
+            return makeUnexpected("Symbolic breakpoint for given symbol, given caseSensitive, and given isRegex already exists"_s);
+    }
+
+    auto& symbolicBreakpoint = m_symbolicBreakpoints.last();
+
+    m_debugger.applyToAllRegisteredCodeBlocks([&] (JSC::CodeBlock* codeBlock) {
+        auto name = getFunctionName(codeBlock);
+
+        if (symbolicBreakpoint.matches(name)) {
+            symbolicBreakpoint.knownMatchingSymbols.add(name);
+
+            JSC::JSLockHolder locker(m_debugger.vm());
+            codeBlock->addBreakpoint(1);
+        }
+    });
+
+    return { };
+}
+
+Protocol::ErrorStringOr<void> InspectorDebuggerAgent::removeSymbolicBreakpoint(const String& symbol, std::optional<bool>&& caseSensitive, std::optional<bool>&& isRegex)
+{
+    SymbolicBreakpoint symbolicBreakpoint;
+    symbolicBreakpoint.symbol = symbol;
+    if (caseSensitive)
+        symbolicBreakpoint.caseSensitive = *caseSensitive;
+    if (isRegex)
+        symbolicBreakpoint.isRegex = *isRegex;
+
+    if (!m_symbolicBreakpoints.removeAll(symbolicBreakpoint))
```

Comment:
Is the `removeAll` just being safe? It seems we already error if there were duplicates.

Reply:
yeah it's just being safe.  we do this for blackbox configs too, so i figured i'd also match it :)